### PR TITLE
improve type annotations in 'docutils.core'

### PR DESCRIPTION
**Pull request by danieleades**

### CODE REVIEW

File: stubs/docutils/docutils/core.pyi [hunk 1]
```diff
@@ -1,3 +1,213 @@
...
+    destination: Output | None
+    destination_class: type[Output]
+    settings: Values | None
+    def __init__(
+        self,
+        reader: Reader[_S] | None = None,
+        parser: Parser | None = None,
+        writer: Writer | None = None,
+        source: Input[_S] | None = None,
+        source_class: type[Input[_S]] = io.FileInput,
```
Comment (JelleZijlstra) [inline stubs/docutils/docutils/core.pyi:30]: There's some mypy errors about this. It's probably difficult to get this right without overloads.

File: stubs/docutils/docutils/core.pyi [hunk 2]
```diff
@@ -1,3 +1,214 @@
...
+        destination_class: type[Output] = io.FileOutput,
+        settings: Values | None = None,
+    ) -> None: ...
+    def set_reader(self, reader_name: str, parser: Parser, parser_name: str) -> None: ...
+    def set_writer(self, writer_name: str) -> None: ...
+    def set_components(self, reader_name: str, parser_name: str, writer_name: str) -> None: ...
+    def setup_option_parser(
+        self,
+        usage: Any | None = None,
+        description: Any | None = None,
```
Comment (srittau) [inline stubs/docutils/docutils/core.pyi:42]: Typed wrong in optparse: #11599

```suggestion
        usage: str | None = None,
        description: str | None = None,
```

File: stubs/docutils/docutils/core.pyi [hunk 3]
```diff
@@ -1,3 +1,214 @@
...
+        settings_spec: SettingsSpec | None = None,
+        config_section: str | None = None,
+        **defaults,
+    ) -> OptionParser: ...
+    def get_settings(
+        self,
+        usage: Any | None = None,
+        description: Any | None = None,
```
Comment (srittau) [inline stubs/docutils/docutils/core.pyi:50]:
```suggestion
        usage: str | None = None,
        description: str | None = None,
```

File: stubs/docutils/docutils/core.pyi [hunk 4]
```diff
@@ -1,3 +1,214 @@
...
+        settings_spec: SettingsSpec | None = None,
+        config_section: str | None = None,
+        **defaults,
+    ) -> Values: ...
+    def process_programmatic_settings(self, settings_spec: SettingsSpec, settings_overrides, config_section: str) -> None: ...
+    def process_command_line(
+        self,
+        argv: list[str] | None = None,
+        usage: Any | None = None,
+        description: Any | None = None,
```
Comment (srittau) [inline stubs/docutils/docutils/core.pyi:60]:
```suggestion
        usage: str | None = None,
        description: str | None = None,
```

File: stubs/docutils/docutils/core.pyi [hunk 5]
```diff
@@ -1,3 +1,214 @@
...
+    ) -> None: ...
+    def set_io(self, source_path: str | None = None, destination_path: str | None = None) -> None: ...
+    def set_source(self, source: Any | None = None, source_path: str | None = None) -> None: ...
+    def set_destination(self, destination: Any | None = None, destination_path: str | None = None) -> None: ...
+    def apply_transforms(self) -> None: ...
+    def publish(
+        self,
+        argv: list[str] | None = None,
+        usage: Any | None = None,
+        description: Any | None = None,
```
Comment (srittau) [inline stubs/docutils/docutils/core.pyi:73]:
```suggestion
        usage: str | None = None,
        description: str | None = None,
```

File: stubs/docutils/docutils/core.pyi [hunk 6]
```diff
@@ -1,3 +1,214 @@
...
+def publish_cmdline(
+    reader: Reader[_S] | None = None,
+    reader_name: str = "standalone",
+    parser: Parser | None = None,
+    parser_name: str = "restructuredtext",
+    writer: Writer | None = None,
+    writer_name: str = "pseudoxml",
+    settings: Values | None = None,
+    settings_spec: SettingsSpec | None = None,
+    settings_overrides: Any | None = None,
```
Comment (srittau) [inline stubs/docutils/docutils/core.pyi:96]: Could probably be typed. Docstring says:

> `settings_overrides`: A dictionary containing application-specific settings defaults that override the defaults of other components. [...]

Without looking at the source, this this probably a `dict`, `Mapping`, or one of the protocols from `_typeshed`. It's okay to leave it as `Incomplete` now, though.

Applies to all the functions below as well.

File: stubs/docutils/docutils/core.pyi [hunk 7]
```diff
@@ -1,3 +1,214 @@
...
+    config_section: str | None = None,
+    enable_exit_status: bool = True,
+    argv: list[str] | None = None,
+    usage: str = default_usage,
+    description: str = default_description,
+) -> Any: ...
```
Comment (srittau) [inline stubs/docutils/docutils/core.pyi:102]: I'd add a comment to this and the other return types below that this is either `str` or `bytes`.

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks, a few remarks below.

**Comment by JelleZijlstra** [comment]
This has a huge merge conflict; do we still need this PR?

**Pull request closed without merging**